src/nlp_expr.jl

```diff
@@ -0,0 +1,825 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+"""
+    NonlinearExpr{V}(head::Symbol, args::Vector{Any})
+    NonlinearExpr{V}(head::Symbol, args::Any...)
```

Comment:
One question: should this be `GenericNonlinearExpr` with `const NonlinearExpr = GenericNonlinearExpr{VariableRef}`?

It think probably, so that we match up with `GenericAffExpr`/`AffExpr` and `GenericQuadExpr`/`QuadExpr`.